consensus/tendermint/core/backlog.go

```diff
@@ -75,59 +75,62 @@ func (c *core) storeBacklog(msg *Message, src types.CommitteeMember) {
 	c.backlogsMu.Lock()
 	defer c.backlogsMu.Unlock()
 
-	backlogPrque := c.backlogs[src]
-	if backlogPrque == nil {
-		backlogPrque = prque.New()
-	}
-	msgRound, errRound := msg.Round()
-	msgHeight, errHeight := msg.Height()
-	if errRound == nil && errHeight == nil {
-		backlogPrque.Push(msg, toPriority(msg.Code, msgRound, msgHeight))
-	}
-
-	c.backlogs[src] = backlogPrque
+	c.backlogs[src] = append(c.backlogs[src], msg)
 }
 
 func (c *core) processBacklog() {
+	var capToLenRatio = 5
 	c.backlogsMu.Lock()
 	defer c.backlogsMu.Unlock()
 
 	for src, backlog := range c.backlogs {
-		if backlog == nil {
-			continue
-		}
-
 		logger := c.logger.New("from", src, "step", c.step)
-		var isFuture bool
-
-		// We stop processing if
-		//   1. backlog is empty
-		//   2. The first message in queue is a future message
-		for !(backlog.Empty() || isFuture) {
-			m, prio := backlog.Pop()
-			msg := m.(*Message)
-			msgRound, _ := msg.Round() // error checking done before push
-			msgHeight, _ := msg.Height()
-
-			// Push back if it's a future message
-			err := c.checkMessage(msgRound, msgHeight, Step(msg.Code))
-			if err != nil {
-				if err == errFutureHeightMessage || err == errFutureRoundMessage || err == errFutureStepMessage {
-					logger.Debug("Stop processing backlog", "msg", msg, "err", err)
-					backlog.Push(msg, prio)
-					isFuture = true
-					break
+
+		initialLen := len(backlog)
+		if initialLen > 0 {
+			// For loop will change the size for backlog therefore we need to keep track of the initial length and
+			// adjust for index change. This is done by keeping track of how many elements have been removed and
+			// subtracting it from the for-loop iterator, since each removed element will cause the index to change for
+			// each element after the removed element.
+			//
+			// If the message is a future height, round or step message then the for-loop will move on to the next
+			// iteration, however, if any other error occurs then the message is removed from the backlog and for-loop
+			// moves to the next iteration.
+			//
+			// If there are no errors when checking the message then the message is sent to the handler via a goroutine
+			// and this message is removed from the backlog.
+			totalElemRemoved := 0
+			for i := 0; i < initialLen; i++ {
+				offset := i - totalElemRemoved
+				curMsg := backlog[offset]
+
+				r, _ := curMsg.Round()
+				h, _ := curMsg.Height()
+				if err := c.checkMessage(r, h, Step(curMsg.Code)); err != nil {
+					if err == errFutureHeightMessage || err == errFutureRoundMessage || err == errFutureStepMessage {
+						logger.Debug("Futrue message in backlog", "msg", curMsg, "err", err)
+						continue
+					}
```

Comment:
I think this could be made a bit neater now.
```suggestion
				err := c.checkMessage(r, h, Step(curMsg.Code))
				if err == errFutureHeightMessage || err == errFutureRoundMessage || err == errFutureStepMessage {
					logger.Debug("Futrue message in backlog", "msg", curMsg, "err", err)
					continue
				}
```